File: Python/leetcode_804.py
'''
Return the number of different transformations
among all words we have.
'''

from typing import List


def unique_morse_representations(words: List[str]) -> int:
    letter_dict = {}
    seen_set: set[str] = set()
    morse = [".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..",
             ".---", "-.-", ".-..", "--",
             "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-",
             "...-", ".--", "-..-", "-.--", "--.."]
    for index, morse_code in enumerate(morse):
        letter_dict[chr(index + 97)] = morse_code
    for word in words:
        morse_string = ''
        word_letter = list(word)
        for letter in word_letter:
            morse_string = morse_string + letter_dict[letter]
        seen_set.add(morse_string)
    return len(seen_set)


print(unique_morse_representations(["gin", "zen", "gig", "msg"]))

File: JavaScript/leetcode_804.js
/**
 * @param {string[]} words
 * @return {number}
 */
var uniqueMorseRepresentations = function (words) {
    let seen = new Set();
    const morse = [
        ".-",
        "-...",
        "-.-.",
        "-..",
        ".",
        "..-.",
        "--.",
        "....",
        "..",
        ".---",
        "-.-",
        ".-..",
        "--",
        "-.",
        "---",
        ".--.",
        "--.-",
        ".-.",
        "...",
        "-",
        "..-",
        "...-",
        ".--",
        "-..-",
        "-.--",
        "--..",
    ];

    for (let index = 0; index < words.length; index++) {
        let morse_string = "";
        let word_letters = [...words[index]];
        for (let index = 0; index < word_letters.length; index++) {
            morse_string = morse_string.concat(
                morse[word_letters[index].charCodeAt() - 97],
            );
        }
        seen.add(morse_string);
    }
    return seen.size;
};
